afr/notify: Fix notify argument for upcall

rafikc30 opened the pull request:
upcall notification expects the upcall data struct
gf_upcall_cache_invalidation as the main argument.
While AFR requires another argument to identify the
child xlator who send the upcall. As of now, afr_notify
was using the upcall structure as the xlator variable
and was calculating the child xlator index value to
a wrong value.

This patch fixes the child index calculation, as well
as tries to have a uniform way to calculate it. This
patch has also moved the handling of notification for
GF_EVENT_TRANSLATOR_OP is the first one to process since
this event doesn't have any meaningful child_xlator
computation.

Change-Id: I7bf03aea53fdedec298869a82efd9b2c6b4e867c
fixes: https://github.com/gluster/glusterfs/issues/3253
Signed-off-by: Mohammed Rafi KC <rafi.kavungal@iternity.com>

rafikc30 committed:
afr/notify: Fix notify argument for upcall

upcall notification expects the upcall data struct
gf_upcall_cache_invalidation as the main argument.
While AFR requires another argument to identify the
child xlator which send the upcall. As of now, afr_notify
was using the upcall strcutre as the xlator variable
and was calculating the child xlator index value to
a wrong value.

This patch fixes the child index calculation, as well
as tries to have a unifrom way to calculate it. This
patch has also moved the hndling of notification for
GF_EVENT_TRANSLATOR_OP as the first one to process since
this event doesnot have any meaningful child_xlator
computation.

Change-Id: I7bf03aea53fdedec298869a82efd9b2c6b4e867c
Signed-off-by: Mohammed Rafi KC <rafi.kavungal@iternity.com>

rafikc30 commented:
/run regression

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/defaults-tmpl.c b/libglusterfs/src/defaults-tmpl.c
index 88a39cde5..f8197a917 100644
--- a/libglusterfs/src/defaults-tmpl.c
+++ b/libglusterfs/src/defaults-tmpl.c
@@ -183,7 +183,8 @@ default_notify(xlator_t *this, int32_t event, void *data, ...)
             xlator_list_t *parent = this->parents;
 
             if (!parent && this->ctx && this->ctx->root)
-                XLATOR_NOTIFY(ret, ((xlator_t*)(this->ctx->root)), event, data, this);
+                XLATOR_NOTIFY(ret, ((xlator_t *)(this->ctx->root)), event, data,
+                              this);
 
             while (parent) {
                 if (parent->xlator->init_succeeded)
diff --git a/xlators/cluster/afr/src/afr-common.c b/xlators/cluster/afr/src/afr-common.c
index 92e775e20..cd20e57ed 100644
--- a/xlators/cluster/afr/src/afr-common.c
+++ b/xlators/cluster/afr/src/afr-common.c
@@ -6228,22 +6228,22 @@ afr_handle_upcall_event(xlator_t *this, struct gf_upcall *upcall)
     }
 }
 
-xlator_t*
+xlator_t *
 afr_get_child_xlator_for_event(int32_t event, void *data, void *data2)
 {
-        xlator_t *child_xlator = NULL;
-        switch (event) {
-                case GF_EVENT_UPCALL:
-                case GF_EVENT_CHILD_PING:
-                        child_xlator = (xlator_t*)data2;
-                        break;
-                case GF_EVENT_TRANSLATOR_OP:
-                        child_xlator = NULL;
-                        break;
-                default:
-                        child_xlator = (xlator_t*)data;
-        }
-        return child_xlator;
+    xlator_t *child_xlator = NULL;
+    switch (event) {
+        case GF_EVENT_UPCALL:
+        case GF_EVENT_CHILD_PING:
+            child_xlator = (xlator_t *)data2;
+            break;
+        case GF_EVENT_TRANSLATOR_OP:
+            child_xlator = NULL;
+            break;
+        default:
+            child_xlator = (xlator_t *)data;
+    }
+    return child_xlator;
 }
 
 int32_t
@@ -6266,7 +6266,6 @@ afr_notify(xlator_t *this, int32_t event, void *data, void *data2)
     int64_t halo_max_latency_msec = 0;
     int64_t child_latency_msec = -1;
 
-
     priv = this->private;
 
     if (!priv)
```

itisravi commented:
>. While AFR requires another argument to identify the child xlator who send the upcall. 

@rafikc30  Is this really the case? It looks like when the event is GF_EVENT_UPCALL,` afr_notify(xlator_t *this, int32_t event, void *data, void *data2)` calls `afr_handle_upcall_event(this, data);`without needing any info about the child xlator which sent the upcall, and `data` as correctly interpreted ad `struct gf_upcall`.

>As of now, afr_notify was using the upcall structure as the xlator variable and was calculating the child xlator index value to a wrong value.

True, but this seems like a harmless no-op as the index is not used like I mentioned above for upcall event. But yes, it is good to skip the calculation.

itisravi commented:
> @rafikc30 Is this really the case? It looks like when the event is GF_EVENT_UPCALL,` afr_notify(xlator_t *this, int32_t event, void *data, void *data2)` calls `afr_handle_upcall_event(this, data);`without needing any info about the child xlator which sent the upcall, and `data` as correctly interpreted ad `struct gf_upcall`.

Ah, you are adding the need for the child xlator in https://github.com/gluster/glusterfs/pull/3255

rafikc30 commented:
> > @rafikc30 Is this really the case? It looks like when the event is GF_EVENT_UPCALL,` afr_notify(xlator_t *this, int32_t event, void *data, void *data2)` calls `afr_handle_upcall_event(this, data);`without needing any info about the child xlator which sent the upcall, and `data` as correctly interpreted ad `struct gf_upcall`.
> 
> Ah, you are adding the need for the child xlator in #3255

@itisravi Yes. I wanted to use the index to identify the Arbiter node.

pranithk commented:
Code looks good to me. Is there a test that can identify the issue you found? If yes could you add it as part of one of the patches?

mykaul commented on xlators/cluster/afr/src/afr-common.c:
```diff
@@ -6228,6 +6228,24 @@ afr_handle_upcall_event(xlator_t *this, struct gf_upcall *upcall)
     }
 }
 
+xlator_t*
```
```suggestion
static xlator_t*
```

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/defaults-tmpl.c b/libglusterfs/src/defaults-tmpl.c
index 88a39cde5..f8197a917 100644
--- a/libglusterfs/src/defaults-tmpl.c
+++ b/libglusterfs/src/defaults-tmpl.c
@@ -183,7 +183,8 @@ default_notify(xlator_t *this, int32_t event, void *data, ...)
             xlator_list_t *parent = this->parents;
 
             if (!parent && this->ctx && this->ctx->root)
-                XLATOR_NOTIFY(ret, ((xlator_t*)(this->ctx->root)), event, data, this);
+                XLATOR_NOTIFY(ret, ((xlator_t *)(this->ctx->root)), event, data,
+                              this);
 
             while (parent) {
                 if (parent->xlator->init_succeeded)
diff --git a/xlators/cluster/afr/src/afr-common.c b/xlators/cluster/afr/src/afr-common.c
index cd6712e19..c24f85fca 100644
--- a/xlators/cluster/afr/src/afr-common.c
+++ b/xlators/cluster/afr/src/afr-common.c
@@ -6251,22 +6251,22 @@ afr_handle_upcall_event(xlator_t *this, struct gf_upcall *upcall, const int idx)
     }
 }
 
-static xlator_t*
+static xlator_t *
 afr_get_child_xlator_for_event(int32_t event, void *data, void *data2)
 {
-        xlator_t *child_xlator = NULL;
-        switch (event) {
-                case GF_EVENT_UPCALL:
-                case GF_EVENT_CHILD_PING:
-                        child_xlator = (xlator_t*)data2;
-                        break;
-                case GF_EVENT_TRANSLATOR_OP:
-                        child_xlator = NULL;
-                        break;
-                default:
-                        child_xlator = (xlator_t*)data;
-        }
-        return child_xlator;
+    xlator_t *child_xlator = NULL;
+    switch (event) {
+        case GF_EVENT_UPCALL:
+        case GF_EVENT_CHILD_PING:
+            child_xlator = (xlator_t *)data2;
+            break;
+        case GF_EVENT_TRANSLATOR_OP:
+            child_xlator = NULL;
+            break;
+        default:
+            child_xlator = (xlator_t *)data;
+    }
+    return child_xlator;
 }
 
 int32_t
@@ -6289,7 +6289,6 @@ afr_notify(xlator_t *this, int32_t event, void *data, void *data2)
     int64_t halo_max_latency_msec = 0;
     int64_t child_latency_msec = -1;
 
-
     priv = this->private;
 
     if (!priv)
```

rafikc30 committed:
glusterd/Arbiter-volgen: Do not add debug xl or user options

While building the server graph, the builder function sometime
decide to skip the xl addition if certain conditions met such
arbiter node will only be added to the last brick in the replica
set. In such cases, there is no need to add debug xl or user
options configured for the xl that we just skipped

Change-Id: I12c574d482bee97335578204d8b1a8292f309c2b
Updates: #3228
Signed-off-by: Mohammed Rafi KC <rafi.kavungal@iternity.com>

rafikc30 committed:
tests/replicate: Add test case for invalid cache update from arbiter

This patch add a test case for issue#3252

Change-Id: I0c407db1028af0e3269ede9514e7063a134d4f69
Updates: #3228
Signed-off-by: Mohammed Rafi KC <rafi.kavungal@iternity.com>

rafikc30 commented:
@pranithk I have added a test case for https://github.com/gluster/glusterfs/issues/3252. I have used delay-gen to create a simple delay in processing an upcall from the arbiter. To support delay-gen for arbiter xlator, I have added a commit.

rafikc30 commented:
/run regression

amarts commented:
looks important. Can this be rebased?

rafikc30 commented:
/run regression
